src/hotspot/share/jfr/instrumentation/jfrResolution.hpp

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_JFR_INSTRUMENTATION_JFRRESOLUTION_HPP
+#define SHARE_JFR_INSTRUMENTATION_JFRRESOLUTION_HPP
+
+#include "memory/allocation.hpp"
+#include "utilities/exceptions.hpp"
+
+class CallInfo;
+class ciKlass;
+class ciMethod;
+class GraphBuilder;
+class Parse;
+
+class JfrResolution : AllStatic {
+ public:
+  static void on_runtime_resolution(const CallInfo & info, TRAPS);
+  static void on_c1_resolution(const GraphBuilder * builder, const ciKlass * holder, const ciMethod * target);
```

Comment:
Should the declarations of `on_c1_resolution` and `on_c2_resolution` be guarded with `COMPILER2_PRESENT` and `COMPILER1_PRESENT` since the method bodies are?

Reply:
Yes. It's been fixed. See:
https://github.com/openjdk/jdk/pull/8680

Reply:
Yes, I see the bodies are guarded. I was referring to the declarations here in `jfrResolution.hpp`. Shouldn't it be something like:
```
#ifdef COMPILER1
  static void on_c1_resolution(const GraphBuilder * builder, const ciKlass * holder, const ciMethod * target);
#endif
#ifdef COMPILER2
  static void on_c2_resolution(const Parse * parse, const ciKlass * holder, const ciMethod * target);
#endif
```
and likewise in `jfr.hpp`:
```
#ifdef COMPILER2
  static void on_resolution(const Parse* parse, const ciKlass* holder, const ciMethod* target);
#endif
#ifdef COMPILER1
  static void on_resolution(const GraphBuilder* builder, const ciKlass* holder, const ciMethod* target);
#endif
```

Reply:
In general, we would like to reduce the amount of conditionalization because it makes the code harder to read and follow. Compilers should not require a definition until a definition is actually needed. Are you seeing compilation errors?

Reply:
No I'm not which surprised me. I agree though, if the compilers don't complain, fewer conditional macros is better.

Reply:
It is because the bodies in the call chain have been conditionalized out. There is no linking because there are no uses.